### PR TITLE
Fix #808: Analytics filter improvements

### LINKED ISSUE #808: Analytics publishing error will result in request to break
### Description:
The analytics filter publishes the request and response path data to analytics server. Error while collecting data to publish will result in an error for the client request. We need to avoid disturbing the request flow whenever analytics publishing flow is broken.

### Affected Product Version:
3.0.1

### OS and other environment details and versions:
- OS:
- Client:
- Env (Docker/K8s):

---

### PR DESCRIPTION
**Pull request by Rajith90**
### Purpose
Error in publishing analytics data will affect the request flow. This PR fixes that issue.

### Issues

Fixes #808 

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 1]
```diff
@@ -25,7 +25,11 @@ public type AnalyticsRequestFilter object {
         if (isAnalyticsEnabled) {
             checkOrSetMessageID(context);
             context.attributes[PROTOCOL_PROPERTY] = caller.protocol;
-            doFilterRequest(request, context);
+            error? err = trap doFilterRequest(request, context);
+            if(err is error) {
```
Comment (hasuniea) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]: There are some formatting issues.

Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
            if (err is error) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 2]
```diff
@@ -37,8 +41,19 @@ public type AnalyticsRequestFilter object {
             if (context.attributes.hasKey(IS_THROTTLE_OUT)) {
                 boolean isThrottleOut = <boolean>context.attributes[IS_THROTTLE_OUT];
                 if (isThrottleOut) {
-                    ThrottleAnalyticsEventDTO eventDto = populateThrottleAnalyticsDTO(context);
-                    eventStream.publish(getEventFromThrottleData(eventDto));
+                    ThrottleAnalyticsEventDTO|error  throttleAnalyticsEventDTO = trap populateThrottleAnalyticsDTO(context);
+                    if(throttleAnalyticsEventDTO is ThrottleAnalyticsEventDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
                    if (throttleAnalyticsEventDTO is ThrottleAnalyticsEventDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 3]
```diff
@@ -37,8 +41,19 @@ public type AnalyticsRequestFilter object {
...
+                        EventDTO|error eventDTO  = trap getEventFromThrottleData(throttleAnalyticsEventDTO);
+                        if(eventDTO is EventDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
                        if (eventDTO is EventDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 4]
```diff
@@ -61,28 +76,47 @@ function doFilterRequest(http:Request request, http:FilterContext context) {
     setRequestAttributesToContext(request, context);
 }
 
-function doFilterFault(http:FilterContext context, error err) {
-    FaultDTO faultDTO = populateFaultAnalyticsDTO(context, err);
-    eventStream.publish(getEventFromFaultData(faultDTO));
+function doFilterFault(http:FilterContext context, string errorMessage) {
+    FaultDTO|error faultDTO = trap populateFaultAnalyticsDTO(context, errorMessage);
+    if(faultDTO is FaultDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
    if (faultDTO is FaultDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 5]
```diff
@@ -61,28 +76,47 @@ function doFilterRequest(http:Request request, http:FilterContext context) {
...
+        EventDTO|error eventDTO = trap getEventFromFaultData(faultDTO);
+        if(eventDTO is EventDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
        if (eventDTO is EventDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 6]
```diff
@@ -61,28 +76,47 @@ function doFilterRequest(http:Request request, http:FilterContext context) {
...
 
 function doFilterResponseData(http:Response response, http:FilterContext context) {
     //Response data publishing
-    RequestResponseExecutionDTO requestResponseExecutionDTO = generateRequestResponseExecutionDataEvent(response,
-        context);
-    EventDTO event = generateEventFromRequestResponseExecutionDTO(requestResponseExecutionDTO);
-    eventStream.publish(event);
+    RequestResponseExecutionDTO|error requestResponseExecutionDTO = trap generateRequestResponseExecutionDataEvent(response,
+            context);
+    if(requestResponseExecutionDTO is RequestResponseExecutionDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
    if (requestResponseExecutionDTO is RequestResponseExecutionDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 7]
```diff
@@ -61,28 +76,47 @@ function doFilterRequest(http:Request request, http:FilterContext context) {
...
+        EventDTO|error event = trap generateEventFromRequestResponseExecutionDTO(requestResponseExecutionDTO);
+        if(event is EventDTO) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
        if (event is EventDTO) {
```

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal [hunk 8]
```diff
@@ -61,28 +76,47 @@ function doFilterRequest(http:Request request, http:FilterContext context) {
...
 
 function doFilterAll(http:Response response, http:FilterContext context) {
-    // TODO: refactor the logic. error does not belong to type any
-    var code = runtime:getInvocationContext().attributes[ERROR_RESPONSE];
-    if (code is ()) {
+    var resp = runtime:getInvocationContext().attributes[ERROR_RESPONSE];
+    if (resp is ()) {
         printDebug(KEY_ANALYTICS_FILTER, "No any faulty analytics events to handle.");
         doFilterResponseData(response, context);
+    } else if(resp is string) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/analytics_request_filter.bal]:
```suggestion
    } else if (resp is string) {
```

**Pull request merged**